src/torch_ucc.cpp

```diff
@@ -718,6 +743,117 @@ ProcessGroupUCC::~ProcessGroupUCC() {
   }
 }
 
+#ifdef USE_CUDA
+// Return CUDA device with ordinal given by input rank.
+at::Device getCUDADeviceForRank(int rank) {
+  TORCH_CHECK(rank >= 0, "Invalid rank ", rank);
+  auto numGPUs = at::cuda::getNumGPUs();
+  int16_t deviceIdx = static_cast<int16_t>(rank % numGPUs);
+  return at::Device(at::DeviceType::CUDA, deviceIdx);
+}
+#endif
+
+void ProcessGroupUCC::runHealthCheck() {
+  // Run health check in a separate thread and wait on CV to handle timeouts.
+  // This design allows us to handle hangs.
+
+  // When size_ is 1, there is no need to do any communication at all.
+  if (size_ == 1) return;
+
+  struct HealthCheckData {
+    std::mutex healthCheckMutex;
+    std::condition_variable healthCheckCv;
+    bool ucxHealthCheckSuccess = false;
+    bool uccHealthCheckSuccess = false;
+    std::exception_ptr healthCheckException;
+  } healthCheckData;
+
+  auto t = std::thread([&healthCheckData, this]() {
+#ifdef USE_CUDA
+    std::list<c10::Device> devices{c10::kCPU};
+    if (at::cuda::is_available()) {
+      devices.emplace_front(getCUDADeviceForRank(rank_));
+    }
+    for (auto device : devices) {
+#else
+    auto device = c10::Device(c10::kCPU);
+#endif
```

Comment:
nit: probably can simplify these few lines to improve readability (Can also remove `#ifdef USE_CUDA` below in lines 824 & 826)

```suggestion
    std::list<c10::Device> devices{c10::kCPU};
#ifdef USE_CUDA
    if (at::cuda::is_available()) {
      devices.emplace_front(getCUDADeviceForRank(rank_));
    }
#endif 
    for (auto device : devices) {
```